src/shims/fs.rs

```diff
@@ -154,27 +166,44 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         this.check_no_isolation("read")?;
 
-        let count = this.read_scalar(count_op)?.to_machine_usize(&*this.tcx)?;
+        let ptr_size = this.pointer_size().bits();
+
+        let count = this
+            .read_scalar(count_op)?
+            .to_machine_usize(&*this.tcx)?
+            .min(1 << (ptr_size - 1));
         // Reading zero bytes should not change `buf`.
         if count == 0 {
             return Ok(0);
         }
         let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf_scalar = this.read_scalar(buf_op)?.not_undef()?;
-
-        // Remove the file handle to avoid borrowing issues.
-        this.remove_handle_and(fd, |mut handle, this| {
-            // Don't use `?` to avoid returning before reinserting the handle.
-            let bytes = this.force_ptr(buf_scalar).and_then(|buf| {
-                // FIXME: Don't use raw methods
-                this.memory
-                    .get_raw_mut(buf.alloc_id)?
-                    .get_bytes_mut(&*this.tcx, buf, Size::from_bytes(count))
-                    .map(|buffer| handle.file.read(buffer))
-            });
-            this.machine.file_handler.handles.insert(fd, handle).unwrap_none();
-            this.try_unwrap_io_result(bytes?.map(|bytes| bytes as i64))
-        })
+        let buf = this.read_scalar(buf_op)?.not_undef()?;
+
+        if let Some(handle) = this.machine.file_handler.handles.get_mut(&fd) {
+            let count = isize::try_from(count).unwrap();
```

Comment:
Please add a comment saying why the `unwrap` is justified.

Reply:
Actually, this one isn't justified, is it? Why did you remove the `ok_or_else` here?

Reply:
I thought that capping `count` would allow us to just unwrap here.

Reply:
No. We capped to the *machine* `isize`, not the *host* `isize`.

We could do that as well though.

Reply:
Which one do you think is better?

Reply:
I am kind of undecided.^^

Yeah I guess just call `min` twice to cap to both, that seems best.

Reply:
So the other size would be `std::mem::sizeof<isize>()` right?

Reply:
Just cap to `isize::MAX as usize`.